### PR TITLE
MicroSALT parameter file creation

### PR DESCRIPTION
**Pull request by ingkebil**
This PR adds config (parameter file) creation for microsalt! It also starts microsalt through cg.

See also: https://github.com/Clinical-Genomics/servers/pull/270

**How to test**:
1. install on stage of the coffee machine: `bash servers/resources/hasta.scilifelab.se/update-cg-stage.sh microsalt`
1. install the new config on stage: `bash servers/resources/hasta.scilifelab.se/update-servers-stage.sh microsalt`
1. activate stage: `us`
1. run following commands: Testing out the config creation `cg workflow microsalt case-config --dry ACC6369A8`
1. `cg workflow microsalt case-config ACC6369A8`
1. `cg workflow microsalt case-config --dry ACC6369`
1. `cg workflow microsalt case-config --dry --project ACC6369`
1. `cg workflow microsalt case-config --project ACC6369`
1. testing out the start command: `cg workflow microsalt run ACC6369 --case-config just-some-case-config --dry`
1. `cg workflow microsalt run ACC6369 --dry`
1. `cg workflow microsalt --project ACC6369`

**Expected outcome**:
4.  
```
[
    {
        "CG_ID_project": "ACC6369",
        "CG_ID_sample": "ACC6369A8",
        "Customer_ID": "cust015",
        "Customer_ID_sample": "ZB1700022A",
        "application_tag": "MWRNXTR003",
        "date_arrival": "0001-01-01 00:00:00",
        "date_libprep": "2019-11-14 00:00:00",
        "date_sequencing": "2019-11-17 13:11:15",
        "method_libprep": "1764:12 - None",
        "method_sequencing": "Not in LIMS",
        "organism": "S. aureus",
        "priority": "research",
        "reference": "AP017922.1"
    }
]
```
5. /home/proj/stage/microbial/queries/ACC6369A8.json should be created. Check with `ls`. Contents should be:
```
[
    {
        "CG_ID_project": "ACC6369",
        "CG_ID_sample": "ACC6369A8",
        "Customer_ID": "cust015",
        "Customer_ID_sample": "ZB1700022A",
        "application_tag": "MWRNXTR003",
        "date_arrival": "0001-01-01 00:00:00",
        "date_libprep": "2019-11-14 00:00:00",
        "date_sequencing": "2019-11-17 13:11:15",
        "method_libprep": "1764:12 - None",
        "method_sequencing": "Not in LIMS",
        "organism": "S. aureus",
        "priority": "research",
        "reference": "AP017922.1"
    }
]
```
6. Should say `Sample ACC6369 not found` and abort.
7. Should output multiple blocks similar to 4.
8.  /home/proj/stage/microbial/queries/ACC6369.json should be created. Check with `ls`. Contents should be the same 7.
9. should print out microsalt command with the alternative case-config
10. should print out microsalt command with the case-config we created, nl /home/proj/stage/microbial/queries/ACC6369.json 
11. should create /home/proj/stage/microbial/queries/ACC6369.json  and start microsalt.

**Review:**
- [x] code approved by @patrikgrenfeldt 
- [x] tests executed by @ingkebil 
- [x] "Merge and deploy" approved by @patrikgrenfeldt 
Thanks for filling in who performed the code review and the test!

This is minor **version bump** because we add new functionality.

### CODE REVIEW

**Comment by ingkebil** [comment]
The config validation has following rules:
- Organism - Fallback based on reference, ‘Other species’ and ‘Comment’. Default to “Unset”.
- Priority = Default to empty string. Weird response. Typically “standard” or “research”. Reference = Defaults to “None”
- Method: Outputted as “1273:23”. Defaults to “Not in LIMS”
- Date: Returns latest == most recent date. Outputted as DT object “YYYY MM DD”.    Defaults to datetime.min

Depending on difficulty both suggested formats and defaults can be fixed in microSALT as opposed to in CG.

**Comment by patrikgrenfeldt** [comment]
Who is the intended audience for your comment? @-mention?

**Comment by ingkebil** [comment]
it's me :)

It's a c/p from a conversation I had with Isak on Slack

**Comment by patrikgrenfeldt** [comment]
Ok I see

Den tors 24 okt. 2019 14:08Kenny Billiau <notifications@github.com> skrev:

**Comment by patrikgrenfeldt** [comment]
coverage increase 🎉

**Comment by ingkebil** [comment]
TC4: 
```
[0|0|0] 1z [kenny.billiau@hasta:~/git/cg] [D_usalt_191119_kb] microsalt 4s ± cg workflow microsalt case-config --dry ACC6369A8
[
    {
        "CG_ID_project": "ACC6369",
        "CG_ID_sample": "ACC6369A8",
        "Customer_ID": "cust015",
        "Customer_ID_sample": "ZB1700022A",
        "application_tag": "MWRNXTR003",
        "date_arrival": "0001-01-01 00:00:00",
        "date_libprep": "2019-11-14 00:00:00",
        "date_sequencing": "2019-11-17 13:11:15",
        "method_libprep": "1764:12",
        "method_sequencing": "Not in LIMS",
        "organism": "S. aureus",
        "priority": "research",
        "reference": "AP017922.1"
    }
]
```

:heavy_check_mark:

**Comment by ingkebil** [comment]
TC5:

```
[0|0|376] 1z [kenny.billiau@hasta:~/git/cg] [D_usalt_191119_kb] microsalt 7s 2 ± cg workflow microsalt case-config ACC6369A8                                                                   
[0|0|376] 1z [kenny.billiau@hasta:~/git/cg] [D_usalt_191119_kb] microsalt 4s ± ll /home/proj/stage/microbial/queries/ACC6369A8.json
-rw-rw-r--+ 1 kenny.billiau hasta-development 531 Jan 18 22:12 /home/proj/stage/microbial/queries/ACC6369A8.json
[0|0|376] 1z [kenny.billiau@hasta:~/git/cg] [D_usalt_191119_kb] microsalt ± cat /home/proj/stage/microbial/queries/ACC6369A8.json
[
    {
        "CG_ID_project": "ACC6369",
        "CG_ID_sample": "ACC6369A8",
        "Customer_ID": "cust015",
        "Customer_ID_sample": "ZB1700022A",
        "application_tag": "MWRNXTR003",
        "date_arrival": "0001-01-01 00:00:00",
        "date_libprep": "2019-11-14 00:00:00",
        "date_sequencing": "2019-11-17 13:11:15",
        "method_libprep": "1764:12",
        "method_sequencing": "Not in LIMS",
        "organism": "S. aureus",
        "priority": "research",
        "reference": "AP017922.1"
    }
]
```
:heavy_check_mark:

**Comment by ingkebil** [comment]
> coverage increase tada

Coincidence. I didn't write tests for anything except the new code. Which is 100% covered tho :)

**Comment by patrikgrenfeldt** [comment]
Is this PR ready for code review?

File: cg/cli/workflow/microsalt/base.py [hunk 1]
```diff
@@ -61,5 +66,80 @@ def link(context: click.Context, order_id: str, sample_id: str):
...
+    microsalt_command = context.obj["usalt"]["binary_path"]
+    command = [microsalt_command]
+
+    case_config_path = case_config
+    if not case_config:
+        queries_path = Path(context.obj["usalt"]["queries_path"])
+        case_config_path = queries_path / project_id
+        case_config_path = case_config_path.with_suffix(".json")
+
+    command.extend(["--parameters", str(case_config_path)])
```
Comment (ingkebil) [inline cg/cli/workflow/microsalt/base.py]:
```suggestion
        command.extend(["--parameters", str(case_config_path)])
```

File: tests/cli/workflow/microsalt/test_microsalt_case_config.py [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+        obj=base_context
+    )
+
+    # THEN command should give us a json dump
+    assert result.exit_code == EXIT_SUCCESS
+    snapshot.assert_match(result.output)
+
+
+def test_sample(cli_runner, base_context, microbial_sample_id, queries_path, snapshot: Snapshot):
+    """Test working dry command for sample"""
```
Comment (ingkebil) [inline tests/cli/workflow/microsalt/test_microsalt_case_config.py]:
```suggestion
    """Test working command for sample"""
```

File: cg/cli/workflow/microsalt/base.py [hunk 3]
```diff
@@ -61,5 +66,80 @@ def link(context: click.Context, order_id: str, sample_id: str):
...
+
+
+@microsalt.command()
+@click.option("-d", "--dry", is_flag=True, help="print command to console")
+@click.option("-c", "--case-config", required=False, help="Optional")
+# @click.option('-p', '--priority', default='low', type=click.Choice(['low', 'normal', 'high']))
+# @click.option('-e', '--email', help='email to send errors to')
+@click.argument("project_id")
+@click.pass_context
+def start(context, dry, case_config, project_id):
```
Comment (ingkebil) [inline cg/cli/workflow/microsalt/base.py]:
```suggestion
def run(context, dry, case_config, project_id):
```

File: tests/cli/workflow/microsalt/test_microsalt_start.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
+""" This file groups all tests related to microsalt start creation """
+
+from cg.cli.workflow.microsalt.base import start
```
Comment (ingkebil) [inline tests/cli/workflow/microsalt/test_microsalt_start.py]:
```suggestion
from cg.cli.workflow.microsalt.base import run
```

File: cg/cli/workflow/microsalt/base.py (hunk 1 above)
Comment (ingkebil) [reply cg/cli/workflow/microsalt/base.py]: Oh, this is a feature: running `run` assumes the case-config file is already there. This options enables testing another case config.

**Comment by ingkebil** [comment]
```
[0|0|163] [kenny.billiau@hasta:~/git/servers] [D_usalt_191119_kb] microsalt(+1/-1) ± cg workflow microsalt case-config --dry ACC6369
Sample ACC6369 not found
Aborted!
```

:heavy_check_mark:

**Comment by ingkebil** [comment]
```
[0|0|162] [kenny.billiau@hasta:~/git/servers] [D_usalt_191119_kb] microsalt(+1/-1) 21s 1 ±  cg workflow microsalt case-config --dry --project 
ACC6369
[
    {
        "CG_ID_project": "ACC6369",
        "CG_ID_sample": "ACC6369A9",
        "Customer_ID": "cust015",
        "Customer_ID_sample": "ZB1700031A",
        "application_tag": "MWRNXTR003",
        "date_arrival": "0001-01-01 00:00:00",
        "date_libprep": "2019-11-14 00:00:00",
        "date_sequencing": "2019-11-22 15:02:16",
        "method_libprep": "1764:12",
        "method_sequencing": "Not in LIMS",
        "organism": "S. aureus",
        "priority": "research",
        "reference": "AP017922.1"
    },
    {
        "CG_ID_project": "ACC6369",
        "CG_ID_sample": "ACC6369A24",
        "Customer_ID": "cust015",
        "Customer_ID_sample": "ZB1700073A",
...
```
:heavy_check_mark:

**Comment by ingkebil** [comment]
5.

```
[0|0|118] [kenny.billiau@hasta:~/git/cg] [D_usalt_191119_kb] microsalt* ± ll /home/proj/stage/microbial/queries/ACC6369.json 
-rw-rw-r--+ 1 kenny.billiau hasta-development 23327 Feb 13 15:16 /home/proj/stage/microbial/queries/ACC6369.json
[0|0|118] [kenny.billiau@hasta:~/git/cg] [D_usalt_191119_kb] microsalt* ± cat /home/proj/stage/microbial/queries/ACC6369.json                 
[
    {
        "CG_ID_project": "ACC6369",
        "CG_ID_sample": "ACC6369A9",
        "Customer_ID": "cust015",
        "Customer_ID_sample": "ZB1700031A",
        "application_tag": "MWRNXTR003",
        "date_arrival": "0001-01-01 00:00:00",
        "date_libprep": "2019-11-14 00:00:00",
        "date_sequencing": "2019-11-22 15:02:16",
        "method_libprep": "1764:12",
        "method_sequencing": "Not in LIMS",
        "organism": "S. aureus",
        "priority": "research",
        "reference": "AP017922.1"
    },
    {
        "CG_ID_project": "ACC6369",
        "CG_ID_sample": "ACC6369A24",
        "Customer_ID": "cust015",
        "Customer_ID_sample": "ZB1700073A",
        "application_tag": "MWRNXTR003",
        "date_arrival": "0001-01-01 00:00:00",
        "date_libprep": "2019-11-14 00:00:00",
        "date_sequencing": "2019-11-22 15:02:16",
        "method_libprep": "1764:12",
        "method_sequencing": "Not in LIMS",
        "organism": "S. aureus",
        "priority": "research",
        "reference": "AP017922.1"
    },
    {
        "CG_ID_project": "ACC6369",
        "CG_ID_sample": "ACC6369A28",
...'
```

:heavy_check_mark:

**Comment by ingkebil** [comment]
TC 9
```
[0|0|111] [kenny.billiau@hasta:~/git/servers] [D_usalt_191119_kb] microsalt(+1/-1) 19s ± cg workflow microsalt run ACC6369 --case-config just-some-case-config --dry
/home/proj/bin/conda/envs/S_microSALT/bin/microSALT --parameters just-some-case-config
```

**Comment by ingkebil** [comment]
TC 10

```
[0|0|111] [kenny.billiau@hasta:~/git/servers] [D_usalt_191119_kb] microsalt(+1/-1) 20s ± cg workflow microsalt run ACC6369 --dry
/home/proj/bin/conda/envs/S_microSALT/bin/microSALT --parameters /home/proj/stage/microbial/queries/ACC6369.json

```

:heavy_check_mark:

**Comment by ingkebil** [comment]
TC 11

```
[0|0|2060] 1z [kenny.billiau@hasta:~/git/cg] [D_usalt_191119_kb] microsalt(+10/-7) 1m50s 148 ± cg workflow microsalt --order ACC6369
ACC6369A9: link FASTQ files
ACC6369A24: link FASTQ files
ACC6369A28: link FASTQ files
ACC6369A1: link FASTQ files
ACC6369A2: link FASTQ files
ACC6369A3: link FASTQ files
ACC6369A4: link FASTQ files
ACC6369A5: link FASTQ files
ACC6369A6: link FASTQ files
ACC6369A7: link FASTQ files
ACC6369A8: link FASTQ files
ACC6369A10: link FASTQ files
ACC6369A11: link FASTQ files
ACC6369A12: link FASTQ files
ACC6369A13: link FASTQ files
ACC6369A14: link FASTQ files
ACC6369A15: link FASTQ files
ACC6369A16: link FASTQ files
ACC6369A17: link FASTQ files
ACC6369A18: link FASTQ files
ACC6369A19: link FASTQ files
ACC6369A20: link FASTQ files
ACC6369A21: link FASTQ files
ACC6369A22: link FASTQ files
ACC6369A23: link FASTQ files
ACC6369A25: link FASTQ files
ACC6369A26: link FASTQ files
ACC6369A27: link FASTQ files
ACC6369A29: link FASTQ files
ACC6369A30: link FASTQ files
ACC6369A31: link FASTQ files
ACC6369A32: link FASTQ files
ACC6369A33: link FASTQ files
ACC6369A34: link FASTQ files
ACC6369A35: link FASTQ files
ACC6369A36: link FASTQ files
ACC6369A37: link FASTQ files
ACC6369A38: link FASTQ files
ACC6369A39: link FASTQ files
ACC6369A40: link FASTQ files
ACC6369A41: link FASTQ files
ACC6369A42: link FASTQ files
ACC6369A43: link FASTQ files
ACC6369A44: link FASTQ files
Starting microSALT! '/bin/true --parameters /home/proj/stage/microbial/queries/ACC6369.json'
```

:checkered_flag:

**Comment by ingkebil** [comment]
The project parameter was changed to `order` after discussion with @emiliaol

**Comment by ingkebil** [comment]
Documentation on how to use the new CLI will be added by @sylvinite . The CLI can then be put into use.

**Pull request merged**